pkg/localstorev2/localstore.go

```diff
@@ -1,25 +1,351 @@
+	// Pins returns all the root references of pinning collections.
+	Pins() ([]swarm.Address, error)
+	// HasPin is a helper which checks if a collection exists with the root
+	// reference passed in.
+	HasPin(swarm.Address) (bool, error)
+}
+
+type memFS struct {
+	afero.Fs
+}
+
+func (m *memFS) Open(path string) (fs.File, error) {
+	return m.Fs.OpenFile(path, os.O_RDWR|os.O_CREATE, 0644)
+}
+
+type dirFS struct {
+	basedir string
+}
+
+func (d *dirFS) Open(path string) (fs.File, error) {
+	return os.OpenFile(filepath.Join(d.basedir, path), os.O_RDWR|os.O_CREATE, 0644)
+}
+
+var sharkyNoOfShards = 32
+
+type closerFn func() error
+
+func (c closerFn) Close() error { return c() }
+
+func CloserFn(closers ...io.Closer) io.Closer {
+	return closerFn(func() error {
+		var err *multierror.Error
+		for _, closer := range closers {
+			err = multierror.Append(err, closer.Close())
+		}
+		return err.ErrorOrNil()
+	})
+}
+
+func initInmemRepository() (*storage.Repository, io.Closer, error) {
+	store, err := leveldbstore.New("", nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed creating inmem levelDB index store: %w", err)
+	}
+
+	sharky, err := sharky.New(
+		&memFS{Fs: afero.NewMemMapFs()},
+		sharkyNoOfShards,
+		swarm.SocMaxChunkSize,
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed creating inmem sharky instance: %w", err)
+	}
+
+	txStore := leveldbstore.NewTxStore(store)
+	txChunkStore := chunkstore.NewTxChunkStore(txStore, sharky)
+
+	return storage.NewRepository(txStore, txChunkStore), CloserFn(store, sharky), nil
+}
+
+// Default options for levelDB.
+var (
+	defaultOpenFilesLimit         = uint64(256)
+	defaultBlockCacheCapacity     = uint64(32 * 1024 * 1024)
+	defaultWriteBufferSize        = uint64(32 * 1024 * 1024)
+	defaultDisableSeeksCompaction = false
+)
+
+func initDiskRepository(basePath string, opts *Options) (*storage.Repository, io.Closer, error) {
+	if opts == nil {
+		opts = &Options{
+			LdbOpenFilesLimit:         defaultOpenFilesLimit,
+			LdbBlockCacheCapacity:     defaultBlockCacheCapacity,
+			LdbWriteBufferSize:        defaultWriteBufferSize,
+			LdbDisableSeeksCompaction: defaultDisableSeeksCompaction,
+		}
+	}
+
+	ldbBasePath := path.Join(basePath, "indexstore")
+
+	if _, err := os.Stat(ldbBasePath); os.IsNotExist(err) {
+		err := os.Mkdir(ldbBasePath, 0775)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	store, err := leveldbstore.New(path.Join(basePath, "indexstore"), &opt.Options{
+		OpenFilesCacheCapacity: int(opts.LdbOpenFilesLimit),
+		BlockCacheCapacity:     int(opts.LdbBlockCacheCapacity),
+		WriteBuffer:            int(opts.LdbWriteBufferSize),
+		DisableSeeksCompaction: opts.LdbDisableSeeksCompaction,
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed creating levelDB index store: %w", err)
+	}
+
+	sharkyBasePath := path.Join(basePath, "sharky")
+
+	if _, err := os.Stat(sharkyBasePath); os.IsNotExist(err) {
+		err := os.Mkdir(sharkyBasePath, 0775)
```

Comment:
The permission is before mask, which means that it would be impossible for user to set permission 0777, as it is limited to 0775 regardless of the mask. I suggest to make directories with 0777.